module-services/service-cellular/ServiceCellular.cpp

```diff
@@ -948,12 +947,22 @@ auto ServiceCellular::receiveSMS(std::string messageNumber) -> bool
     std::string messageRawBody;
     UTF8 receivedNumber;
     const auto &cmd = at::factory(at::AT::QCMGR);
-    auto ret        = channel->cmd(cmd + messageNumber, cmd.getTimeout());
-    if (!ret) {
-        LOG_ERROR("!!!! Could not read text message !!!!");
-        retVal = false;
+    at::Result ret;
+    auto qcmgrRetries = 0;
+
+    while (qcmgrRetries < at::AtCmdMaxRetries) {
+        ret = channel->cmd(cmd + messageNumber, cmd.getTimeout());
+        if (!ret) {
```

Comment:
should we retry for each error?

Reply:
I've been wondering about that, too. We could warp the execution of every command with a retry just in case.

Reply:
@alekrudnik tbh we should rather have a separate "algorithm piece of code (not embedded strictly into the `ServiceCellular.cpp` which would handle receiving messages as a whole - both short, and concatenated ones.

@niepiekm what is the exact error reason when we trigger the delay? do you know that `channel->cmd()` awaits for "data or error" for selected time?

Reply:
As I mentioned [above](https://github.com/mudita/MuditaOS/pull/3512/files#r791508562) the delay is just for retry of a command. Respective command execution time is handleded by the `channel-cmd()` itself per command.

Reply:
Adding a default delay in cmd() seems to not be a good decision. Lets add some encapsulation/policy in place so it would be easily changeable for specific commands.